File: 210330_BOJ_3190.cpp
#include <iostream>
#include <memory.h>
#include <vector>
#include <deque>

using namespace std;

int main() {

	int N;
	cin >> N;

	bool isApple[101][101];
	memset(isApple, false, sizeof(isApple));
	int x, y;
	int K;
	cin >> K;
	for (int i = 0; i < K; ++i) {
		cin >> x >> y;
		isApple[x][y] = true;
	}

	char command;
	int L;
	cin >> L;
	vector<pair<int, char>> orders;
	for (int i = 0; i < L; ++i) {
		cin >> x >> command;
		orders.push_back({ x, command });
	}
	orders.push_back({0, 'N'});

	int dir[4][2]{ {-1, 0}, {0, 1}, {1, 0}, {0, -1} };
	int head = 1;
	x = y = 1;
	int oIndex = 0;
	deque<pair<int, int>> dq;
	dq.push_back({ 1, 1 });

	int seconds = 1;
	while (1) {

		int px = x + dir[head][0];
		int py = y + dir[head][1];

		if (px <= 0 || px > N || py <= 0 || py > N)
			break;

		bool isSnakeBody = false;
		for (auto it : dq) {
			if (it.first == px && it.second == py) {
				isSnakeBody = true;
				break;
			}
		}

		if (isSnakeBody)
			break;

		x = px, y = py;

		dq.push_front({ x, y });
		if (isApple[x][y]) {
			isApple[x][y] = false;
		}
		else {
			dq.pop_back();
		}

		if (seconds == orders[oIndex].first) {
			if (orders[oIndex].second == 'L') {
				head--;
				if (head < 0)
					head = 3;
				head = head % 4;
			}
			else {
				head = (head + 1) % 4;
			}
			oIndex++;
		}

		seconds++;
	}

	cout << seconds << "\n";
	return 0;
}

File: 210330_BOJ_3190.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;

public class Main {

    static class II {
        int x;
        int y;

        public II(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    static class IC {
        int s;
        char d;

        public IC(int s, char d) {
            this.s = s;
            this.d = d;
        }
    }

    public static void main(String[] args) throws IOException {

        int N = 0;
        int m = 0;
        int L = 0;
        boolean[][] isApple = new boolean[101][101];
        IC[] order = new IC[101];
        int[][] dir = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());

        m = Integer.parseInt(br.readLine());
        for(int i = 0; i < m; ++i) {
            String c = br.readLine();
            String[] arr = c.split(" ");
            isApple[Integer.parseInt(arr[0])][Integer.parseInt(arr[1])] = true;
        }

        L = Integer.parseInt(br.readLine());
        for(int i = 0; i < L; ++i) {
            String c = br.readLine();
            String[] arr = c.split(" ");
            order[i] = new IC(Integer.parseInt(arr[0]), arr[1].charAt(0));
        }
        order[L] = new IC(0, 'N');

        ArrayDeque<II> dq = new ArrayDeque<>();
        dq.offerLast(new II(1, 1));

        int oIndex = 0;
        int x = 1;
        int y = 1;
        int head = 1;
        int seconds = 1;
        while(true) {

            int px = x + dir[head][0];
            int py = y + dir[head][1];

            if (px <= 0 || px > N || py <= 0 || py > N)
                break;

            boolean isSnakeBody = false;
            for (II p : dq) {
                if (p.x == px && p.y == py) {
                    isSnakeBody = true;
                    break;
                }
            }

            if (isSnakeBody)
                break;

            x = px;
            y = py;

            dq.offerFirst(new II(x, y));
            if (isApple[x][y]) {
                isApple[x][y] = false;
            }
            else {
                dq.pollLast();
            }

            if (order[oIndex].s == seconds) {
                if (order[oIndex].d == 'L') {
                    head--;
                    if (head == -1)
                        head = 3;
                    head = head % 4;
                }
                else if (order[oIndex].d == 'D') {
                    head = (head + 1) % 4;
                }
                oIndex++;
            }

            seconds++;
        }

        System.out.println(seconds);
    }
}